File: solutions/0002_add-two-numbers.py
# Problem: https://leetcode.com/problems/add-two-numbers
# Runtime: 46 ms

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        out = ListNode()
        curr = out
        while True:
            n1 = l1.val if l1 else 0
            n2 = l2.val if l2 else 0
            res = n1 + n2 + carry
            curr.val = res % 10
            carry = res // 10

            #
            l1 = l1.next if l1 else l1
            l2 = l2.next if l2 else l2
            if (l1 == None and l2 == None):
                if (carry != 0):
                    curr.next = ListNode(carry)
                break
            
            curr.next = ListNode()
            curr = curr.next

        return out

File: solutions/0002_add-two-numbers.java
// Problem: https://leetcode.com/problems/add-two-numbers

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int carry = 0;
        ListNode out = null;
        ListNode curr = null;
        while (l1 != null || l2 != null) {
            int l1Val = l1 == null ? 0 : l1.val;
            int l2Val = l2 == null ? 0 : l2.val;
            int val = l1Val + l2Val + carry;
            if (out == null) {
                out = new ListNode(val % 10);
                curr = out;
            } else {
                curr.next = new ListNode(val % 10);
                curr = curr.next;
            }
            carry = val / 10;
            if (l1 != null)
                l1 = l1.next;
            if (l2 != null)
                l2 = l2.next;
        }
        if (carry != 0) {
            curr.next = new ListNode(carry);
        }
        return out;
    }
}